### ISSUE TITLE
analyze: tracking issue for simple_buffer

### ISSUE THREAD

**Issue by spernsteiner** [immunant/c2rust#1103]
`simple_buffer` is a simplified buffer implementation (similar to `Vec<u8>`) designed to mimic lighttpd's `buffer` module.

* [ ] borrowck panic: Missing origin params on `Rvalue::Cast` output.  This is triggered by `malloc(...) as *mut buffer`.  `buffer` has one origin param for its internal data pointer, but the `borrowck::LTy` computed for the cast has no origin params.
* [ ] unlower panic: Missing entry for `Rvalue::Ref` arising from `let ref mut fresh = ...;`.  
* [ ] borrowck: `return malloc(...)` removes `UNIQUE` from the output pointer because Polonius reports an error.  
  * Needs more investigation to understand the precise cause of this error
* [ ] borrowck: Drops `UNIQUE` due to overlapping borrows around `realloc`:
  ```C
  buf->data = realloc(buf->data, ...);
  ```
  ```Rust
  let old = (*buf).data;
  let fresh = &mut (*buf).data;
  *fresh = realloc(old, ...);
  ```
* [ ] borrowck: Drops `UNIQUE` due to two-phase-like behavior in `buffer_realloc(buf, (*buf).cap * 2)`.
  * [ ] Check whether lighttpd buffer actually uses this pattern
* [ ] mir_op: `malloc`/`realloc` calls aren't recognized as rewritable via `SafeMalloc`/`SafeRealloc`.
  * Needs more investigation
* [ ] `derive(Copy)` is left on `struct buffer` even after rewriting the `data` pointer to use `Box`.
* [ ] rewrite: Many bad interactions between `Option` (`NULL`) and `Result` (`DynOwned`/`FREE`) rewrites.
  * [ ] Develop test cases for individual issues

**Comment by spernsteiner** [comment]
We can now analyze, rewrite, compile, and run `simple_buffer` after patching it with workarounds for various `c2rust-analyze` issues.  As we fix the issues tracked here, we can remove the workarounds until `simple_buffer` works unmodified.

**Comment by spernsteiner** [comment]
I've implemented fully interprocedural pointee type analysis (on a dev branch - no PR yet), and it now produces correct pointee types for the `data` field and for all `malloc`/`free`/`realloc` calls.  Next I'm planning to look at the borrowck `Rvalue::Cast` issue.
